src/java/org/apache/cassandra/index/sai/disk/vector/VectorMemtableIndex.java

```diff
@@ -374,18 +375,25 @@ protected PrimaryKey computeNext()
 
     private class KeyFilteringBits implements Bits
     {
-        private final Set<PrimaryKey> results;
+        private final List<PrimaryKey> results;
 
         public KeyFilteringBits(List<PrimaryKey> results)
         {
-            this.results = new HashSet<>(results);
+            this.results = results;
         }
 
         @Override
         public boolean get(int i)
         {
-            var pk = graph.keysFromOrdinal(i);
-            return results.stream().anyMatch(pk::contains);
+            // pks is not ordered. If it were, we could use the result from binarySearch to search a sublist of
+            // results on subsequent searches. It's not common to have multiple PKs for an ordinal, though.
+            var pks = graph.keysFromOrdinal(i);
+            for (var pk : pks)
+                // We use this class when we have more than maxBruteForceRows results, so binary search
+                // should always be faster than linear search. We could possibly
+                if (Collections.binarySearch(results, pk) >= 0)
+                    return true;
+            return false;
```

Comment:
@jbellis - looks like this is the only place where we call `hashCode()` on a PK. I can see the value in `O(1)` look ups here, but notably, we were only iterating over the HashSet. This implementation is strictly better, but isn't `O(m * log(n))` where m is number of pks. In a memtable, we have a fully materialized PK, so `loadDeferred()` is not expensive, but depending on the limit, it might be more expensive to insert all of the PKs into the hash set instead of just doing the `topk` binary searches. What is your opinion?

Reply:
how about if we leave this alone, and implement hashcode just for the non-deferred / memtable case?  (i.e. if primaryKeySupplier != null then throw)

Reply:
I forgot that just because we're searching the memtable doesn't mean all of our search PKs are loaded.

I think the next best solution is to hash the token and the clustering columns. Hashing the partition key doesn't seem valuable. Also, we know at Primary Key creation time whether it has clustering columns, so we can use that to determine whether hashCode requires loading the key.